### PR TITLE
Implemented content-fairness ranking, resolved issue in novelty, cleaned up code and repo

### PR DESCRIPTION
**Pull request by jannikjw**
- Implemented content fairness ranking
- Changed novelty calculation (based on probabilities and on full dataset not only slate)
- extended experiments.py with more optional arguments
- changed user pairing to be based on user clusters
- deleted old measurement files
- fixed issues in cosine similarity (value below zero would mess up novelty metric -> all prediction scores are shifted to be above 0)

### CODE REVIEW

File: src/scoring_functions.py [hunk 1]
```diff
@@ -28,4 +33,40 @@ def entropy(predicted_user_profiles, predicted_item_attributes):
...
     re_ranked_scores = predicted_scores + alpha * entropy
-    return re_ranked_scores
\ No newline at end of file
+    assert (re_ranked_scores >= 0).all(), "Some scores are negative."
+
+    return re_ranked_scores
+
+
+def content_fairness(predicted_user_profiles, predicted_item_attributes):
+    slate_size = 10
```
Comment (madisonthantu) [inline src/scoring_functions.py:65]: if there's any chance that we reduce/change the slate size maybe consider making this a hyperparameter

File: src/scoring_functions.py [hunk 2]
```diff
@@ -28,4 +33,40 @@ def entropy(predicted_user_profiles, predicted_item_attributes):
...
     
     re_ranked_scores = predicted_scores + alpha * entropy
-    return re_ranked_scores
\ No newline at end of file
+    assert (re_ranked_scores >= 0).all(), "Some scores are negative."
+
+    return re_ranked_scores
+
+
+def content_fairness(predicted_user_profiles, predicted_item_attributes):
```
Comment (madisonthantu) [inline src/scoring_functions.py:41]: I'm having trouble following this 😎

File: wrapper/metrics/evaluation_metrics.py [hunk 3]
```diff
@@ -240,4 +245,59 @@ def measure(self, recommender):
...
+        """
+        Measures the proportion of relevant items (i.e., those users interacted with) falling
+        within the top k ranked items shown..
+        Parameters
+        ------------
+            recommender: :class:`~models.recommender.BaseRecommender`
+                Model that inherits from
+                :class:`~models.recommender.BaseRecommender`.
+        """
+        if self.k >= recommender.num_items_per_iter:
```
Comment (madisonthantu) [inline wrapper/metrics/evaluation_metrics.py:281]: move to init

File: wrapper/metrics/evaluation_metrics.py [hunk 4]
```diff
@@ -240,4 +245,59 @@ def measure(self, recommender):
...
+
+        interactions = recommender.interactions
+        if interactions.size == 0:
+            self.observe(None)  # no interactions yet
+            return
+
+        else:
+            shown_item_scores = np.take(recommender.predicted_scores.value, recommender.items_shown)
+            shown_item_ranks = np.argsort(shown_item_scores, axis=1)
+            top_k_items = np.take(recommender.items_shown, shown_item_ranks[:, self.k :])
```
Comment (madisonthantu) [inline wrapper/metrics/evaluation_metrics.py]: not sure but should it be `shown_item_ranks[:,recommender.items_per_iter-self.k:]`

File: clustering_experiments.ipynb [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
...46NnXISSTRDZvJTKyR/BMrRexyh9UBMt00rH2t8iKi7Cp3ySro13o/oaqJz4YRR338nVjZ0FGjuKR8zzsCsiglVybD5U/p5ufJGx0yYQ8g0py7MCOH+On6D35J1qjetZUHXOnuHjweW9fVw+cKqf17R7ypb55AAZI0Wb1sx4x2TslXa8ko+htpEElb6njt7M2Ho7L24oT8BXJAFZEfjLM+VWBo0dRe5+LsKVC4Ns2FNuXJcrmqRTEAp5adUSNBf7vv+/+bVh2w6TqeOGsj7f2634DGUa+Wjp0RRWvHy77lJ+2/4cew92Mx+uVnJz1ZlUqaWcjIDspEENMNs9Ap/kRBVKIkqzDHJW37/RTdkmxtirCYguNuaa+HnrE1wUOIWNuQN9bl+wNKyDn8IC2vQsTYUGfvtAnNzBxG9FErjxzEqGiAJqp87Hzq3kkVdjFDSLglba99JZUNz9994HMHWSe5/FN/LcPq7WwWtoagiCBQeLDPx2iynDndR4FBYMcVJ4OE5ke54Ft1aRzZrs3lQq9xcEmHeuj/Ez/j1yKo+H4xI6X/jCF/j973/PFVdcMdDnc0yMHTuWV1/tu7RukEGOhKA4cdfMIN2yEiMfx14xBk/9AhzB0aSbX8fqw5fiRDH1NI7QBIqpA+S7eipm7BVj8TSciniMuU92Ocik8K1ktVYi2Y3Y5ABBxwQcSrhXxVUqZ/C35yIs29TzcNvbVmB4VYDPnZWgavrNRLfdh5ZpBwSc1afgHXI67Wt+A4CqiDizy4ltuw+AXGQ7zqqpZNvX9z6v4BiKicbSZ9YydG28i9pT/x+yPYCo9F+Zs62xdGNu7CgwqlZlV0vvaa9zpwmYrU/hH30h6QM9EaX03qV86eLP8uMHLeJpA7sqcMEcP0um+1Ckkyeio+mlaarWqIbXJdDo3M/eQpRx1ZO4aK6fZ9YmyBctnDaRC+b4qG/I8VByQ9kY872LeKDrbwDYRDuL/ecf83lMGOrE65RIHuYTNW2ks5f4OUR7TMPnklizozyXq65SpSVdpFMv7Tfe3nd0cGSNnVLdXIkXV+e4dfQcijadLYVS3zQZkXOc00ns83LamJ7fyRBbJV+vvZikURJZXsmJV+5Z31iM0qzFiOhpHouv50zvBJaltpe5Bb+ZrFnELiqc4hrKt5ofwMAkYxTwSc5+O6cfXvp+kTqP/7srRlHvESWa
```
Comment (madisonthantu) [inline clustering_experiments.ipynb:240]: no output shown for this cell onwards

**Review by madisonthantu — COMMENTED** [review]
Left some comments but overall looks ~~beautiful~~

File: src/scoring_functions.py (hunk 1 above)
Comment (jannikjw) [reply src/scoring_functions.py:65]: Unfortunately, you cannot pass parameters to the score_fn. It is obviously bad practice but I would say it's totally fine for our project.

File: src/scoring_functions.py (hunk 2 above)
Comment (jannikjw) [reply src/scoring_functions.py:41]: Fair, I added documentation - does it make more sense now?

File: wrapper/metrics/evaluation_metrics.py (hunk 3 above)
Comment (jannikjw) [reply wrapper/metrics/evaluation_metrics.py:281]: the recommender is not know in init

**Comment by jannikjw** [comment]
I am right, you are wrong

**Pull request merged**